### PR TITLE
feat: Emit management canister idl during build when imported by Motoko canister

### PR DESCRIPTION
**Pull request by kentosugama**
# Description

When Motoko canisters import the management canister
```
import management "ic:aaaaa-aa";
```
it would be ideal if the interface is automatically provided by `dfx`. 

The execution team is planning on adding the idl to the replica, but until then it can be hard coded in `dfx` and swapped out later on.

# How Has This Been Tested?

```
@test "build supports auto-generated idl for management canister imports in motoko" {
    install_asset motoko_management
    dfx_start
    dfx canister create --all
    assert_command dfx build
    dfx deploy
    assert_command dfx canister call e2e_project_backend rand
}
```
```
import Management "ic:aaaaa-aa";

actor {
  public func rand() : async Blob {
    await Management.raw_rand();
  };
};

```

And the recursive version of the test.
```
import Rand "dependency";

actor {
  public func rand() : async Blob {
    await Rand.rand();
  };
};
```
`dependency.mo`
```
import Management "ic:aaaaa-aa";

module Rand {
  public func rand() : async Blob {
    await Management.raw_rand();
  };
};
```

# Checklist:

- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [x] I have made corresponding changes to the documentation.
- [ ] Different subnets may have different management canisters

### CODE REVIEW

**Comment by chenyan-dfinity** [comment]
> The execution team is planning on adding the idl to the replica

Not going to happen. The difficulty is that we don't have a way to route aaaaa-aa, so it's not easy to get this endpoint.

**Comment by kentosugama** [comment]
I spoke with @dsarlis this morning in the dx working group and he mentioned he was looking into building the feature

**Comment by chenyan-dfinity** [comment]
That would be great!

**Comment by kentosugama** [comment]
@ericswanson-dfinity Hi Eric. Not quite sure where the appropriate place for documenting this feature would be. I'm considering `dfx build --help`.

**Comment by ericswanson-dfinity** [comment]
> @ericswanson-dfinity Hi Eric. Not quite sure where the appropriate place for documenting this feature would be. I'm considering `dfx build --help`.

How about here? https://github.com/dfinity/sdk/blob/master/docs/cli-reference/dfx-build.md

File: src/dfx/src/lib/builders/motoko.rs [hunk 1]
```diff
@@ -132,6 +138,16 @@ impl CanisterBuilder for MotokoBuilder {
...
             .with_context(|| format!("Failed to create {}.", idl_dir_path.to_string_lossy()))?;
 
+        // If the management canister is being imported, emit the candid file.
+        if get_imports(cache.as_ref(), input_path)?
+            .contains(&MotokoImport::Ic("aaaaa-aa".to_string()))
+        {
+            let management_idl_path = idl_dir_path.join("aaaaa-aa.did");
+            std::fs::write(&management_idl_path, management_idl()?).with_context(|| {
+                format!("Failed to write {}.", management_idl_path.to_string_lossy())
+            })?;
```
Comment (ericswanson-dfinity) [inline src/dfx/src/lib/builders/motoko.rs:148]:
```suggestion
            dfx_core::fs::write(&management_idl_path, management_idl()?)?;
```

File: src/dfx/src/util/assets.rs [hunk 2]
```diff
@@ -107,3 +107,205 @@ pub fn bitcoin_wasm(logger: &slog::Logger) -> DfxResult<Vec<u8>> {
         bail!("Failed to find bitcoin canister archive entry");
     }
 }
+
```
Comment (ericswanson-dfinity) [inline src/dfx/src/util/assets.rs:110]: Would you please add a comment here containing a URL (somewhere in https://github.com/dfinity/ic) to where this came from?

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]

File: src/dfx/src/util/assets.rs (hunk 2 above)
Comment (kentosugama) [reply src/dfx/src/util/assets.rs:110]: Would this be okay? https://github.com/dfinity/interface-spec/blob/master/spec/_attachments/ic.did

Comment (kentosugama) [reply src/dfx/src/util/assets.rs:110]: Follow up: it seems that the `did` file is written manually (and verified by the CI) so I believe `interface-spec` would be the closest thing to a source of truth there is.

Comment (ericswanson-dfinity) [reply src/dfx/src/util/assets.rs:110]: Sounds good.  Basically trying to find out: where should a maintainer look in order to find updates to this?

Comment (kentosugama) [reply src/dfx/src/util/assets.rs:110]: The interface spec would be the place to get it from. I just forwarded you a thread on Slack

File: src/dfx/src/lib/builders/motoko.rs [hunk 3]
```diff
@@ -132,6 +138,14 @@ impl CanisterBuilder for MotokoBuilder {
         std::fs::create_dir_all(idl_dir_path)
             .with_context(|| format!("Failed to create {}.", idl_dir_path.to_string_lossy()))?;
 
+        // If the management canister is being imported, emit the candid file.
+        if get_imports(cache.as_ref(), input_path)?
```
Comment (ericswanson-dfinity) [inline src/dfx/src/lib/builders/motoko.rs:142]: Will this emit the management canister candid file if the import comes not from the "main" motoko file, but from a .mo file that it imports?

The best way to make sure would be to add an e2e test.

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]

File: src/dfx/src/lib/builders/motoko.rs (hunk 3 above)
Comment (kentosugama) [reply src/dfx/src/lib/builders/motoko.rs:137]: Apologies, getting back to this now. Will address this issue.

File: src/dfx/src/lib/builders/motoko.rs [hunk 4]
```diff
@@ -38,59 +39,59 @@ impl MotokoBuilder {
     }
 }
 
+#[context("Failed to find imports for canister at '{}'.", file.to_string_lossy())]
```
Comment (ericswanson-dfinity) [inline src/dfx/src/lib/builders/motoko.rs:42]:
```suggestion
#[context("Failed to find imports for canister at '{}'.", file.display())]
```

File: src/dfx/src/lib/builders/motoko.rs [hunk 5]
```diff
@@ -38,59 +39,59 @@ impl MotokoBuilder {
...
+fn get_imports(cache: &dyn Cache, file: &Path) -> DfxResult<BTreeSet<MotokoImport>> {
+    #[context("Failed recursive dependency detection at {}.", file.to_string_lossy())]
```
Comment (ericswanson-dfinity) [inline src/dfx/src/lib/builders/motoko.rs:44]:
```suggestion
    #[context("Failed recursive dependency detection at {}.", file.display())]
```

File: src/dfx/src/lib/builders/motoko.rs [hunk 6]
```diff
@@ -38,59 +39,59 @@ impl MotokoBuilder {
...
+    fn get_imports_recursive(
+        cache: &dyn Cache,
+        file: &Path,
```
Comment (ericswanson-dfinity) [inline src/dfx/src/lib/builders/motoko.rs:47]: In the previous code, this `file` parameter wasn't shadowing an outer `file` parameter.  This seems error-prone.  What would you think about moving this inner function out?

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]

File: src/dfx/src/lib/builders/motoko.rs (hunk 6 above)
Comment (kentosugama) [reply src/dfx/src/lib/builders/motoko.rs:47]: I refactored by passing in the Motoko canister info to the outer function

**Review by ericswanson-dfinity — APPROVED** [review]

**Pull request merged**